lib/logdna/client.rb

```diff
@@ -1,34 +1,52 @@
 # frozen_string_literal: true
 
+require "etc"
 require "net/http"
 require "socket"
 require "json"
 require "concurrent"
 require "date"
+require 'securerandom'
 
 module Logdna
+  Message = Struct.new(:source, :size)
+
   class Client
     def initialize(request, uri, opts)
       @uri = uri
 
       # NOTE: buffer is in memory
       @buffer = []
-      @buffer_byte_size = 0
-
-      @side_messages = []
 
       @lock = Mutex.new
-      @side_message_lock = Mutex.new
-      @flush_limit = opts[:flush_size] || Resources::FLUSH_BYTE_LIMIT
+      @flush_threshold = opts[:flush_size] || Resources::FLUSH_BYTE_LIMIT
+      @flush_limit = 2
+
       @flush_interval = opts[:flush_interval] || Resources::FLUSH_INTERVAL
-      @flush_scheduled = false
-      @exception_flag = false
 
       @request = request
       @retry_timeout = opts[:retry_timeout] || Resources::RETRY_TIMEOUT
 
       @internal_logger = Logger.new(STDOUT)
       @internal_logger.level = Logger::DEBUG
+
+      @work_thread_pool = Concurrent::FixedThreadPool.new(Etc.nprocessors)
+      @request_thread_pool = Concurrent::FixedThreadPool.new(Resources::MAX_CONCURRENT_REQUESTS)
```

Comment:
Should this be marked as daemons, so they won't potentially block the process termination (eg: weird crashes on jruby)?

Reply:
The options are correct as far as I can tell. `auto_terminate` (default: true) should allow the thread pools to shutdown when the program tries to exit